### PR TITLE
Fix: Error creating file manifest for LungMAP project (#3982)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Connected issue: #3982

### Author

- [x] Target branch is `develop`
- [x] Source branch is `issues/<github-handle>/<issue#>-<slug>`
- [x] PR title references the connected issue
- [x] PR title matches<sup>1</sup> title of connected issue         <sub>or comment in PR explains why they're different</sub>
- [x] Title of at least one commit references connected issue
- [x] PR is connected to issue via Zenhub 
- [x] PR description links to connected issue
- [x] Added `partial` label to PR                                    <sub>or this PR completely resolves the connected issue</sub>

<sup>1</sup> when the issue title describes a problem, the PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain                <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title                            <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfix connected to the issue          <sub>or there is no temporary hotfix for the connected issue on the `prod` branch</sub>

### Author (requirements)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (rebasing, integration test)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after rejection)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `demo` or `no demo`
- [x] Commented on connected issue about demo expectations          <sub>or labelled connected issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Ticket is in *Review requested* column
- [x] Requested review from peer reviewer
- [x] Assigned PR to peer reviewer

### Peer reviewer (after rejection)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Peer reviewer (after approval)

- [x] `N reviews` label on PR reflects prior peer reviews with changes requested
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or connected issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Branch pushed to GitLab and added `sandbox` label             <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                                       <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox`                     <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issue to Merged column
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [ ] Shortened the PR chain                                        <sub>or this PR is not labeled `chain`</sub>
- [ ] Pushed merge commit to GitLab                                 <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from GitHub and GitLab
- [ ] Build passes on GitLab

### Operator (reindex) 

- [ ] Deleted unreferenced indices in `dev`                         <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [ ] Started reindex in `dev`                                      <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures                     <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment                      <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: src/azul/service/manifest_service.py [hunk 1]
```diff
@@ -916,8 +916,7 @@ def manifest_content_hash(self) -> int:
 
     def file_name(self, object_key, base_name: Optional[str] = None) -> str:
         if base_name:
-            file_name_prefix = unicodedata.normalize('NFKD', base_name)
-            file_name_prefix = re.sub(r'[^\w ,.@%&\-_()\\[\]/{}]', '_', file_name_prefix).strip()
```
Comment (hannes-ucsc) [inline src/azul/service/manifest_service.py:920]: The file name still needs to be sanitized like this. Only the tag value must be base64-encoded.

File: src/azul/service/manifest_service.py [hunk 2]
```diff
@@ -916,8 +916,7 @@ def manifest_content_hash(self) -> int:
...
+            file_name_prefix = base64.urlsafe_b64encode(base_name.encode()).decode()
```
Comment (hannes-ucsc) [inline src/azul/service/manifest_service.py:919]: Ideally we wouldn't need to decode again and use the bytes as the tag value. If tag value must be a string, decode ASCII.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/service/manifest_service.py (hunk 2 above)
Comment (danielsotirhos) [reply src/azul/service/manifest_service.py]: Tags values must be strings.

```
...
2022-04-15 14:19:26,105 INFO    MainThread: Tagging object 'manifests/e9309ecd-3f7b-56b1-b54f-72a203c802a7.64610d37-a501-5b3a-a7e8-2b7f10e37144.tsv' with {'TagSet': [{'Key': 'azul_file_name', 'Value': b'TW91c2UgTWVsYW5vbWEgMjAyMi0wNC0xNSAxNC4xOS50c3Y='}]}
2022-04-15 14:19:26,689 DEBUG   MainThread: Tearing Down Data

Error
Traceback (most recent call last):
...
  File "/Users/daniel/repo/azul2/src/azul/service/manifest_service.py", line 1061, in write
    self.storage.put_object_tagging(object_key, tagging)
  File "/Users/daniel/repo/azul2/src/azul/service/storage_service.py", line 212, in put_object_tagging
    self.client.put_object_tagging(Bucket=self.bucket_name,
  File "/Users/daniel/repo/azul2/.venv/lib/python3.8/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/daniel/repo/azul2/.venv/lib/python3.8/site-packages/botocore/client.py", line 648, in _make_api_call
    request_dict = self._convert_to_request_dict(
  File "/Users/daniel/repo/azul2/.venv/lib/python3.8/site-packages/botocore/client.py", line 696, in _convert_to_request_dict
    request_dict = self._serializer.serialize_to_request(
  File "/Users/daniel/repo/azul2/.venv/lib/python3.8/site-packages/botocore/validate.py", line 293, in serialize_to_request
    raise ParamValidationError(report=report.generate_report())
botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid type for parameter Tagging.TagSet[0].Value, value: b'TW91c2UgTWVsYW5vbWEgMjAyMi0wNC0xNSAxNC4xOS50c3Y=', type: <class 'bytes'>, valid types: <class 'str'>
```

File: src/azul/service/manifest_service.py [hunk 3]
```diff
@@ -933,7 +934,11 @@ def file_name(self, object_key, base_name: Optional[str] = None) -> str:
         return file_name
 
     def tagging(self, file_name: Optional[str]) -> Optional[Mapping[str, str]]:
-        return None if file_name is None else {self.service.file_name_tag: file_name}
+        if file_name is None:
+            return None
+        else:
+            encoded_file_name = base64.urlsafe_b64encode(file_name.encode())
+            return {self.service.file_name_tag: encoded_file_name.decode('ascii')}
```
Comment (hannes-ucsc) [inline src/azul/service/manifest_service.py:941]: There should be total anti-symmetry. Decoding should be the exact opposite of encoding. The fact that `'asci'` only occurs once is a red flag.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Review by achave11 — APPROVED** [review]
LGTM!

**Pull request merged**